mods/screwdriver/init.lua

```diff
@@ -19,6 +11,72 @@ screwdriver.rotate_simple = function(pos, node, user, mode, new_param2)
 	end
 end
 
+-- For attached wallmounted nodes: returns true if rotation is valid
+local function check_attached_node(pos, rotation)
+	local d = minetest.wallmounted_to_dir(rotation)
+	local p2 = vector.add(pos, d)
+	local n = minetest.get_node(p2).name
+	local def2 = minetest.registered_nodes[n]
+	if def2 and not def2.walkable then
+		return false
+	end
+	return true
+end
+
+screwdriver.rotate = {}
+
+local facedir_tbl = {
+	[screwdriver.ROTATE_FACE] = {
+		[0] = 1, [1] = 2, [2] = 3, [3] = 0,
+		[4] = 5, [5] = 6, [6] = 7, [7] = 4,
+		[8] = 9, [9] = 10, [10] = 11, [11] = 8,
+		[12] = 13, [13] = 14, [14] = 15, [15] = 12,
+		[16] = 17, [17] = 18, [18] = 19, [19] = 16,
+		[20] = 21, [21] = 22, [22] = 23, [23] = 20,
+	},
+	[screwdriver.ROTATE_AXIS] = {
+		[0] = 4, [1] = 4, [2] = 4, [3] = 4,
+		[4] = 8, [5] = 8, [6] = 8, [7] = 8,
+		[8] = 12, [9] = 12, [10] = 12, [11] = 12,
+		[12] = 16, [13] = 16, [14] = 16, [15] = 16,
+		[16] = 20, [17] = 20, [18] = 20, [19] = 20,
+		[20] = 0, [21] = 0, [22] = 0, [23] = 0,
+	},
+}
+
+screwdriver.rotate.facedir = function(pos, node, mode)
+	local rotation = node.param2 % 32 -- get first 5 bits
+	local other = node.param2 - rotation
+	rotation = facedir_tbl[mode][rotation] or 0
+	return rotation + other
+end
+
+screwdriver.rotate.colorfacedir = screwdriver.rotate.facedir
+
+local wallmounted_tbl = {
+	[screwdriver.ROTATE_FACE] = {[2] = 5, [3] = 4, [4] = 2, [5] = 3, [1] = 0, [0] = 1},
+	[screwdriver.ROTATE_AXIS] = {[2] = 5, [3] = 4, [4] = 2, [5] = 1, [1] = 0, [0] = 3}
+}
+
+screwdriver.rotate.wallmounted = function(pos, node, mode)
+	local rotation = node.param2 % 8 -- get first 3 bits
+	local other = node.param2 - rotation
+	rotation = wallmounted_tbl[mode][rotation] or 0
+	if minetest.get_item_group(node.name, "attached_node") ~= 0 then
+		-- find an acceptable orientation
+		for i = 1, 5 do
```

Comment:
If the new position is unacceptable (the sign would pop off the wall) the loop tries the next rotation five times and stops if a rotation is acceptable. The loop repeats five times because rhe rotation was already changed once before the loop.